### ISSUE TITLE
search/filter commit(s)

### ISSUE THREAD

**Issue by extrawurst** [extrawurst/gitui#1791]
We want to substring / fuzzy search for 
* commit sha (or parts of it), 
* authors fields
* commit message
* filenames appearing in diff
* content of diffs

fixes #449, #429, #1790

Note: **Fuzzy** does not mean realtime in this case. It means we do not need perfect matching, mostly relying on `fuzzy_matcher`

**Comment by AmmarAbouZor** [comment]
#### SHA

I think we can at first separate using sha from the rest because it's one to one search (each sha will give one commit only).
For that  I would suggest a "jump to commit" function which opens an input box with a keybinding to past from the system clipboard then when it's applied the commit will be selected if found or some message can be shown

#### Filter/Fuzzy Matches

I think there is a trade off we need to make here...
- Live fuzzy matches is more intuitive and easier to use but it needs all the data to be loaded in memory so it will not give misleading information.
- A "classic" Filter function: Filter popup with the wanted options >> apply the filter to show only the commits that meet the filter criteria. This approach isn't as intuitive as the fuzzy matches one but it comes with less overhead since we don't need to have all the data in memory and the filtering will be applied only once and not with each keystroke.

**Comment by bewlay** [comment]
Listing all commits that modify a specific file would be very useful (maybe this is covered by diff filtering).

**Comment by bewlay** [comment]
Indeed, I wasn't aware of that feature. Very nice.

**Comment by extrawurst** [comment]
@AmmarAbouZor your comment lead me to split out a separate issue for this particular use case: https://github.com/extrawurst/gitui/issues/1799

**Comment by Rodrigodd** [comment]
Recently (a month ago, actually) I have opened the PR #1753, that finish the work done by #506 and #672. That PR implements searching by sha, author, message and tag, including controls for each one, but not in a fuzzy way.

I am not sure if you @extrawurst missed it, or it just isn't relevant for this issue.
